### PR TITLE
Add a way to decouple the implementation and the declaration of a TyCtxt method.

### PR DESCRIPTION
**Pull request by oli-obk**
properly addresses https://github.com/rust-lang/rust/pull/115819

accepted MCP: https://github.com/rust-lang/compiler-team/issues/395

### CODE REVIEW

File: src/librustdoc/core.rs [hunk 1]
```diff
@@ -286,7 +286,7 @@ pub(crate) fn create_config(
                 let body = hir.body(hir.body_owned_by(def_id));
                 debug!("visiting body for {def_id:?}");
                 EmitIgnoredResolutionErrors::new(tcx).visit_body(body);
-                (rustc_interface::DEFAULT_QUERY_PROVIDERS.typeck)(tcx, def_id)
+                (rustc_interface::DEFAULT_QUERY_PROVIDERS.0.typeck)(tcx, def_id)
```
Comment (WaffleLapkin) [inline src/librustdoc/core.rs:289]: Would it maybe be better to make `DEFAULT_QUERY_PROVIDERS` have a `struct DefaultQueryProviders { queries: ..., hooks: ... }`? I don't really like `.0` here...

~~wishing we had anon structs 😅~~

**Review by WaffleLapkin — APPROVED** [review]
r=me, modulo the nit which I'm not sure makes sense
@rustbot author

File: src/librustdoc/core.rs (hunk 1 above)
Comment (oli-obk) [reply src/librustdoc/core.rs]: I also merged the extern providers into that new struct, but I can drop that commit.

**Review by WaffleLapkin — APPROVED** [review]
r=me with the CI failure fixed

File: compiler/rustc_middle/src/hooks/mod.rs [hunk 2]
```diff
@@ -0,0 +1,65 @@
+use crate::mir;
+use crate::query::TyCtxtAt;
+use crate::ty::{Ty, TyCtxt};
+use rustc_span::DUMMY_SP;
+
+macro_rules! declare_hooks {
```
Comment (RalfJung) [inline compiler/rustc_middle/src/hooks/mod.rs:6]: Would be good to have some docs in this file that explain hat a "hook" is and how it differs from a "query".

File: compiler/rustc_middle/src/query/mod.rs [hunk 3]
```diff
@@ -1100,16 +1100,6 @@ rustc_queries! {
         desc { "destructuring type level constant"}
     }
 
-    /// Tries to destructure an `mir::Const` ADT or array into its variant index
-    /// and its field values. This should only be used for pretty printing.
-    query try_destructure_mir_constant_for_diagnostics(
-        key: (mir::ConstValue<'tcx>, Ty<'tcx>)
```
Comment (RalfJung) [inline compiler/rustc_middle/src/query/mod.rs:1106]: So can we now remove the faulty StableHash implementation(s) that were previously needed to make this work as a query?

**Comment by oli-obk** [comment]
We need to either backport this PR or just apply https://github.com/rust-lang/rust/pull/115819 to beta without applying it to master

**Comment by apiraino** [comment]
backport declined in favor of #115819 as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-09-28/near/393729489)

@rustbot label -beta-nominated

File: compiler/rustc_middle/src/query/mod.rs (hunk 3 above)
Comment (WaffleLapkin) [reply compiler/rustc_middle/src/query/mod.rs:1106]: What "faulty StableHash implementationt(s)" are you thinking about?

Comment (RalfJung) [reply compiler/rustc_middle/src/query/mod.rs:1106]: I'm quoting things @oli-obk said, I don't actually understand the underlying issue here. Also [see Zulip](https://rust-lang.zulipchat.com/#narrow/stream/146212-t-compiler.2Fconst-eval/topic/HashStable.20for.20AllocId.3F).

Comment (WaffleLapkin) [reply compiler/rustc_middle/src/query/mod.rs:1106]: Read the zulip, still don't get it 👍🏻 

Also the "hash eq implies eq" is very scary...

**Pull request merged**